framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -272,106 +273,206 @@ protected boolean areEqualInHierarchy(
+     *
+     * <p>Roughly, the algorithm works as follows (assuming {@code canBeCovariant} is false):
+     *
+     * <ul>
+     *   <li>If inside is a bound type, recur on the explicit bound.
+     *   <li>If one of outside's bounds is itself a bound type, recur on that bound.
+     *   <li>Otherwise, return {@code outside.lower <: inside && inside <: outside.upper}.
+     * </ul>
+     *
+     * @param inside a possibly-contained type
+     * @param outside a possibly-containing type
+     * @param canBeCovariant whether or not type arguments are allowed to be covariant
+     * @return true if inside is contained by outside, or if canBeCovariant == true and {@code
+     *     inside <: outside}
+     */
+    protected boolean isContainedByBoundType(
+            AnnotatedTypeMirror inside, BoundType outside, boolean canBeCovariant) {
+        if (BoundType.isBoundType(inside)) {
+            BoundType insideBoundType = new BoundType(inside);
+            if (!insideBoundType.hasExplicitLowerBound && !outside.hasExplicitLowerBound) {
+                return (canBeCovariant || isSubtype(outside.lower, insideBoundType.lower))
+                        && isContainedByBoundType(insideBoundType.upper, outside, canBeCovariant);
+            } else if (outside.hasExplicitLowerBound
+                    || TypesUtils.isObject(outside.upper.getUnderlyingType())) {
+                return (canBeCovariant || isSubtype(insideBoundType.upper, outside.upper))
+                        && isContainedByBoundType(insideBoundType.lower, outside, canBeCovariant);
             }
-            outsideUpperBound = ((AnnotatedWildcardType) outsideUpperBound).getExtendsBound();
         }
 
-        AnnotatedTypeMirror castedInside =
-                AnnotatedTypes.castedAsSuper(inside.atypeFactory, inside, outsideUpperBound);
-        if (!isSubtypeCaching(castedInside, outsideUpperBound)) {
-            return false;
+        if (BoundType.isBoundType(outside.upper)) {
+            BoundType outsideUpper = new BoundType(outside.upper);
+            Set<AnnotationMirror> setA =
+                    AnnotatedTypes.findEffectiveLowerBoundAnnotations(
+                            qualifierHierarchy, outsideUpper.lower);
+            Set<AnnotationMirror> setB =
+                    AnnotatedTypes.findEffectiveLowerBoundAnnotations(
+                            qualifierHierarchy, outside.lower);
+            Set<? extends AnnotationMirror> glb =
+                    qualifierHierarchy.greatestLowerBounds(setA, setB);
+            addToLowestBound(outside.lower, glb);
+            return isContainedByBoundType(inside, outsideUpper, canBeCovariant);
+        } else if (BoundType.isBoundType(outside.lower)) {
+            BoundType outsideLower = new BoundType(outside.lower);
+            return isContainedByBoundType(inside, outsideLower, canBeCovariant);
+        } else {
+            if (canBeCovariant) {
+                if (outside.hasExplicitLowerBound) {
```

Comment:
Is this `if` test an optimization to save resources compared to the `return ... && ...` statement on line 435, or is it logically necessary?